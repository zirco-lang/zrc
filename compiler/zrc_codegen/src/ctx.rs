//! Structures used internally within the code generator for context and state
//! management

use inkwell::{
    builder::Builder,
    context::Context,
    debug_info::{DICompileUnit, DILexicalBlock, DebugInfoBuilder},
    module::Module,
    targets::TargetMachine,
    values::FunctionValue,
};
use zrc_utils::line_finder::LineLookup;

use crate::CgScope;

/// Trait for any context with at least the fields of [`CompilationUnitCtx`]
#[allow(dead_code)]
pub trait AsCompilationUnitCtx<'ctx: 'a, 'a> {
    /// Convert some specific context to a [`CompilationUnitCtx`]
    fn as_unit_ctx(&self) -> CompilationUnitCtx<'ctx, 'a>;

    /// The LLVM context
    fn ctx(&self) -> &'ctx Context {
        self.as_unit_ctx().ctx
    }
    /// The LLVM target machine
    fn target_machine(&self) -> &'a TargetMachine {
        self.as_unit_ctx().target_machine
    }
    /// The LLVM builder used to build instructions
    fn builder(&self) -> &'a Builder<'ctx> {
        self.as_unit_ctx().builder
    }
    /// The lookup for lines in the source file
    fn line_lookup(&self) -> &'a LineLookup {
        self.as_unit_ctx().line_lookup
    }
    /// The LLVM builder for debug info
    fn dbg_builder(&self) -> &'a DebugInfoBuilder<'ctx> {
        self.as_unit_ctx().dbg_builder
    }
    /// The LLVM compile unit for debug info
    fn compilation_unit(&self) -> &'a DICompileUnit<'ctx> {
        self.as_unit_ctx().compilation_unit
    }
    /// The LLVM module we are building in
    fn module(&self) -> &'a Module<'ctx> {
        self.as_unit_ctx().module
    }
}

/// LLVM structures common to a single compilation unit (file)
#[derive(Debug, Clone, Copy)]
pub struct CompilationUnitCtx<'ctx, 'a> {
    /// The LLVM context
    pub ctx: &'ctx Context,
    /// The LLVM target machine
    pub target_machine: &'a TargetMachine,
    /// The LLVM builder used to build instructions
    pub builder: &'a Builder<'ctx>,
    /// The lookup for lines in the source file
    pub line_lookup: &'a LineLookup,
    /// The LLVM builder for debug info
    pub dbg_builder: &'a DebugInfoBuilder<'ctx>,
    /// The LLVM compile unit for debug info
    pub compilation_unit: &'a DICompileUnit<'ctx>,
    /// The LLVM module we are building in
    pub module: &'a Module<'ctx>,
}
impl<'ctx, 'a> AsCompilationUnitCtx<'ctx, 'a> for CompilationUnitCtx<'ctx, 'a> {
    fn as_unit_ctx(&self) -> Self {
        *self
    }
}

/// Common LLVM structures passed onto most code generation functions
#[derive(Debug, Clone, Copy)]
pub struct FunctionCtx<'ctx, 'a> {
    // == FROM CompilationUnitCtx ==
    /// The LLVM context
    pub ctx: &'ctx Context,
    /// The LLVM target machine
    pub target_machine: &'a TargetMachine,
    /// The LLVM builder used to build instructions
    pub builder: &'a Builder<'ctx>,
    /// The lookup for lines in the source file
    pub line_lookup: &'a LineLookup,
    /// The LLVM builder for debug info
    pub dbg_builder: &'a DebugInfoBuilder<'ctx>,
    /// The LLVM compile unit for debug info
    pub compilation_unit: &'a DICompileUnit<'ctx>,
    /// The LLVM module we are building in
    pub module: &'a Module<'ctx>,

    /// The LLVM function we are building in
    pub fn_value: FunctionValue<'ctx>,
}
impl<'ctx, 'a> AsCompilationUnitCtx<'ctx, 'a> for FunctionCtx<'ctx, 'a> {
    fn as_unit_ctx(&self) -> CompilationUnitCtx<'ctx, 'a> {
        CompilationUnitCtx {
            ctx: self.ctx,
            target_machine: self.target_machine,
            builder: self.builder,
            line_lookup: self.line_lookup,
            dbg_builder: self.dbg_builder,
            compilation_unit: self.compilation_unit,
            module: self.module,
        }
    }
}
impl<'ctx, 'a> FunctionCtx<'ctx, 'a> {
    /// Create a function context from a [`CompilationUnitCtx`] and
    /// [`FunctionValue`]
    pub const fn from_unit_and_fn(
        unit: CompilationUnitCtx<'ctx, 'a>,
        fn_value: FunctionValue<'ctx>,
    ) -> Self {
        Self {
            ctx: unit.ctx,
            target_machine: unit.target_machine,
            builder: unit.builder,
            line_lookup: unit.line_lookup,
            dbg_builder: unit.dbg_builder,
            compilation_unit: unit.compilation_unit,
            module: unit.module,
            fn_value,
        }
    }
}

/// An extension of [`FunctionCtx`] containing values found in an individual
/// lexical block (except `bb`, as this is mutated often)
#[derive(Debug, Clone, Copy)]
pub struct BlockCtx<'ctx, 'input, 'a> {
    // == FROM CompilationUnitCtx ==
    /// The LLVM context
    pub ctx: &'ctx Context,
    /// The LLVM target machine
    pub target_machine: &'a TargetMachine,
    /// The LLVM builder used to build instructions
    pub builder: &'a Builder<'ctx>,
    /// The lookup for lines in the source file
    pub line_lookup: &'a LineLookup,
    /// The LLVM builder for debug info
    pub dbg_builder: &'a DebugInfoBuilder<'ctx>,
    /// The LLVM compile unit for debug info
    pub compilation_unit: &'a DICompileUnit<'ctx>,
    /// The LLVM module we are building in
    pub module: &'a Module<'ctx>,

    // == FROM FunctionCtx ==
    /// The LLVM function we are building in
    pub fn_value: FunctionValue<'ctx>,

    /// The code generation type/value scope this block lives in
    pub scope: &'a CgScope<'input, 'ctx>,
    /// The LLVM [`DILexicalBlock`] we're currently in
    pub dbg_scope: DILexicalBlock<'ctx>,
}
impl<'ctx, 'a> AsCompilationUnitCtx<'ctx, 'a> for BlockCtx<'ctx, '_, 'a> {
    fn as_unit_ctx(&self) -> CompilationUnitCtx<'ctx, 'a> {
        CompilationUnitCtx {
            ctx: self.ctx,
            target_machine: self.target_machine,
            builder: self.builder,
            line_lookup: self.line_lookup,
            dbg_builder: self.dbg_builder,
            compilation_unit: self.compilation_unit,
            module: self.module,
        }
    }
}
impl<'ctx, 'input, 'a> BlockCtx<'ctx, 'input, 'a> {
    /// Create a new [`BlockCtx`] from a [`FunctionCtx`] and needed block
    /// scoping parameters
    pub const fn new(
        function_ctx: FunctionCtx<'ctx, 'a>,
        scope: &'a CgScope<'input, 'ctx>,
        dbg_scope: DILexicalBlock<'ctx>,
    ) -> Self {
        Self {
            ctx: function_ctx.ctx,
            target_machine: function_ctx.target_machine,
            builder: function_ctx.builder,
            line_lookup: function_ctx.line_lookup,
            dbg_builder: function_ctx.dbg_builder,
            compilation_unit: function_ctx.compilation_unit,
            module: function_ctx.module,
            fn_value: function_ctx.fn_value,
            scope,
            dbg_scope,
        }
    }
}
