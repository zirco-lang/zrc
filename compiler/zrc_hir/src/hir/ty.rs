use std::fmt::Display;

#[derive(PartialEq, Debug, Clone)]
pub enum Type {
    Identifier(String),
}

impl Display for Type {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Type::Identifier(i) => write!(f, "{}", i),
        }
    }
}
