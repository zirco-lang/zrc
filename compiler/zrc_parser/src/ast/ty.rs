//! Type representation for the Zirco AST
//!
//! The main thing within this module you will need is the [`Type`] struct. It
//! contains all the different type kinds in Zirco.

use std::fmt::Display;

use derive_more::Display;
use zrc_utils::{
    span::{Span, Spannable, Spanned},
    spanned,
};

/// A valid Zirco AST type
#[derive(PartialEq, Eq, Debug, Clone, Display)]
#[display("{_0}")]
pub struct Type<'input>(pub Spanned<TypeKind<'input>>);

/// The key-value pairs of a struct
#[derive(PartialEq, Eq, Debug, Clone)]
#[allow(clippy::type_complexity)]
pub struct KeyTypeMapping<'input>(pub Spanned<Vec<Spanned<(Spanned<&'input str>, Type<'input>)>>>);
impl Display for KeyTypeMapping<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for (i, member) in self.0.value().iter().enumerate() {
            write!(f, "{}: {}", member.value().0.value(), member.value().1)?;
            if i < self.0.value().len() - 1 {
                write!(f, ", ")?;
            }
        }
        Ok(())
    }
}

/// A valid Zirco AST type
#[derive(PartialEq, Eq, Debug, Clone, Display)]
pub enum TypeKind<'input> {
    /// An identifier, such as `i32`
    #[display("{_0}")]
    Identifier(&'input str),
    /// `*T`
    #[display("*{_0}")]
    Ptr(Box<Type<'input>>),
    /// A direct struct type
    #[display("struct {{ {_0} }}")]
    Struct(KeyTypeMapping<'input>),
    /// A direct union type
    #[display("union {{ {_0} }}")]
    Union(KeyTypeMapping<'input>),
    /// A tagged union type
    #[display("enum {{ {_0} }}")]
    Enum(KeyTypeMapping<'input>),
}

// AST builder. We are able to infer the spans of many based on the start of
// their leftmost and the end of their rightmost operands.
#[allow(missing_docs)]
#[allow(clippy::missing_docs_in_private_items)]
#[allow(clippy::should_implement_trait)]
impl<'input> Type<'input> {
    #[must_use]
    pub fn build_ident(ident: Spanned<&'input str>) -> Self {
        let span = ident.span();
        Self(spanned!(
            span.start(),
            TypeKind::Identifier(ident.into_value()),
            ident.end(),
            span.file_name()
        ))
    }

    #[must_use]
    pub fn build_struct_from_contents(span: Span, keys: KeyTypeMapping<'input>) -> Self {
        Self(TypeKind::Struct(keys).in_span(span))
    }

    #[must_use]
    pub fn build_union_from_contents(span: Span, keys: KeyTypeMapping<'input>) -> Self {
        Self(TypeKind::Union(keys).in_span(span))
    }

    #[must_use]
    pub fn build_enum_from_contents(span: Span, keys: KeyTypeMapping<'input>) -> Self {
        Self(TypeKind::Enum(keys).in_span(span))
    }

    #[must_use]
    pub fn build_ptr(span: Span, ty: Self) -> Self {
        Self(TypeKind::Ptr(Box::new(ty)).in_span(span))
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn types_stringify_to_their_canonical_form() {
        // A list of sample types in "canonical form."
        // These are parsed then stringified again, and tested for equality.

        let test_cases = vec![
            "i32",
            "*i32",
            "struct { a: i32, b: i32 }",
            "union { a: i32, b: i32 }",
            "enum { Eight: i8, Sixteen: i16 }",
        ];

        for input in test_cases {
            assert_eq!(
                crate::parser::parse_type(input, "<test>")
                    .expect("test cases should have parsed correctly")
                    .to_string(),
                input
            );
        }
    }

    #[test]
    fn parenthesized_types_parse_correctly() {
        // Parenthesized types should parse and normalize to their canonical form
        // (without unnecessary parentheses)

        let test_cases = vec![
            ("*(i32)", "*i32"),
            ("**(i32)", "**i32"),
            ("*(struct { a: i32 })", "*struct { a: i32 }"),
            ("*(union { a: i32, b: i32 })", "*union { a: i32, b: i32 }"),
            // Nested parentheses
            ("*((i32))", "*i32"),
            ("*((struct { a: i32 }))", "*struct { a: i32 }"),
            // Parenthesized types in struct fields
            ("struct { x: (i32) }", "struct { x: i32 }"),
        ];

        for (input, expected) in test_cases {
            let parsed = crate::parser::parse_type(input, "<test>")
                .unwrap_or_else(|_| panic!("Failed to parse: {input}"));
            assert_eq!(
                parsed.to_string(),
                expected,
                "Input: {input} should normalize to: {expected}"
            );
        }
    }

    #[test]
    fn type_aliases_support_parenthesized_types() {
        // Type aliases should support parenthesized types including nested parentheses
        let test_cases = vec!["type x = (i32);", "type y = ((i32));", "type z = (*(i32));"];

        for input in test_cases {
            let result = crate::parser::parse_program(input, "<test>");
            assert!(result.is_ok(), "Failed to parse type alias: {input}");
        }
    }
}
