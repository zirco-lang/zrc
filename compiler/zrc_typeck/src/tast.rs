//! Type-checked AST for the Zirco programming language
//!
//! This is similar to the [standard AST](zrc_parser::ast), but with type
//! information attached to it.
//!
//! Note that we only attach source locations to code that would exist after code generation,
//! meaning inlined types do not get a span.

pub mod expr;
pub mod stmt;
pub mod ty;
