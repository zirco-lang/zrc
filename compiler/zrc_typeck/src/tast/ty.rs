//! Type representation for the Zirco [TAST](super)

use std::{collections::HashMap, fmt::Display};

use crate::typeck::BlockReturnType;

/// The possible Zirco types
#[derive(PartialEq, Debug, Clone)]
pub enum Type {
    // WHENEVER ADDING NEW PRIMITIVES HERE, ADD THEM TO THE TYPE SCOPE IN
    // [`zrc_typeck::typeck::Scope::default`].
    /// `i8`
    I8,
    /// `u8`
    U8,
    /// `i16`
    I16,
    /// `u16`
    U16,
    /// `i32`
    I32,
    /// `u32`
    U32,
    /// `i64`
    I64,
    /// `u64`
    U64,
    /// `void`, only producible by calling a void function (`fn()`)
    Void,
    /// `bool`
    Bool, /* TODO: need an "any Int" type that implicitly casts to all int types but becomes
           * i32 when assigned to a value */
    /// `*T`
    Ptr(Box<Type>),
    /// `fn(A, B) -> T`
    Fn(Vec<Type>, Box<BlockReturnType>),
    /// Struct type literals
    Struct(HashMap<String, Type>),
}

impl Display for Type {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self.clone() {
            Self::I8 => write!(f, "i8"),
            Self::U8 => write!(f, "u8"),
            Self::I16 => write!(f, "i16"),
            Self::U16 => write!(f, "u16"),
            Self::I32 => write!(f, "i32"),
            Self::U32 => write!(f, "u32"),
            Self::I64 => write!(f, "i64"),
            Self::U64 => write!(f, "u64"),
            Self::Bool => write!(f, "bool"),
            Self::Ptr(t) => write!(f, "*({t})"),
            Self::Void => write!(f, "void"),
            Self::Fn(args, brt) => write!(
                f,
                "(fn({}){})",
                args.iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", "),
                match *brt {
                    BlockReturnType::Return(ret) => format!(" -> {ret}"),
                    BlockReturnType::Void => String::new(),
                }
            ),
            Self::Struct(fields) => write!(
                f,
                "(struct {{ {} }})",
                fields
                    .iter()
                    .map(|(k, v)| format!("{k}: {v}"))
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
        }
    }
}

impl Type {
    /// Returns `true` if this is an integer type like [`Type::I8`].
    #[must_use]
    pub const fn is_integer(&self) -> bool {
        use Type::{I16, I32, I64, I8, U16, U32, U64, U8};
        matches!(self, I8 | U8 | I16 | U16 | I32 | U32 | I64 | U64)
    }

    /// Returns `true` if this is a signed integer type like [`Type::I8`].
    #[must_use]
    pub const fn is_signed_integer(&self) -> bool {
        use Type::{I16, I32, I64, I8};
        matches!(self, I8 | I16 | I32 | I64)
    }

    /// Returns `true` if this is an unsigned integer type like [`Type::U8`].
    #[must_use]
    pub const fn is_unsigned_integer(&self) -> bool {
        use Type::{U16, U32, U64, U8};
        matches!(self, U8 | U16 | U32 | U64)
    }
}
