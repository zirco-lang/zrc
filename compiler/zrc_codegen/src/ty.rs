//! Resolution from [`Type`] instances to LLVM types.
//!
//! This module contains functions to convert Zirco types
//! ([`zrc_typeck::tast::ty::Type`]) to their corresponding
//! LLVM types ([`inkwell::types::AnyTypeEnum`]).
//!
//! The main function is [`llvm_type`], which takes a Zirco type and
//! returns the corresponding LLVM type. There are also helper functions
//! for creating pointers and function types.

use inkwell::{
    AddressSpace,
    debug_info::{AsDIScope, DIBasicType, DIDerivedType, DISubroutineType, DIType},
    types::{
        AnyType, AnyTypeEnum, BasicMetadataTypeEnum, BasicType, BasicTypeEnum, FunctionType,
        IntType, PointerType,
    },
};
use zrc_typeck::tast::ty::{Fn, Type};

use crate::ctx::AsCompilationUnitCtx;

/// Create a pointer to an [`AnyTypeEnum`] instance.
///
/// # Panics
/// Panics if `ty` is [`AnyTypeEnum::VoidType`].
fn create_ptr<'ctx: 'a, 'a>(
    ctx: &impl AsCompilationUnitCtx<'ctx, 'a>,
    ty: AnyTypeEnum<'ctx>,
    dbg_ty: DIType<'ctx>,
) -> (PointerType<'ctx>, DIDerivedType<'ctx>) {
    (
        match ty {
            AnyTypeEnum::ArrayType(x) => x.ptr_type(AddressSpace::default()),
            AnyTypeEnum::FloatType(x) => x.ptr_type(AddressSpace::default()),
            AnyTypeEnum::IntType(x) => x.ptr_type(AddressSpace::default()),
            AnyTypeEnum::PointerType(x) => x.ptr_type(AddressSpace::default()),
            AnyTypeEnum::StructType(x) => x.ptr_type(AddressSpace::default()),
            AnyTypeEnum::VectorType(x) => x.ptr_type(AddressSpace::default()),
            AnyTypeEnum::VoidType(_) => panic!("cannot create a pointer to void"),
            AnyTypeEnum::FunctionType(x) => x.ptr_type(AddressSpace::default()),
        },
        ctx.dbg_builder().create_pointer_type(
            &ty.to_string(),
            dbg_ty,
            0,
            0,
            AddressSpace::default(),
        ),
    )
}
/// Create a function pointer from a prototype.
///
/// Returns a [`DIBasicType`] because for some reason [`DISubroutineType`] can't
/// be cast to a [`DIType`]. Use .1 when doing ``.create_function``, use .2 for
/// other uses
///
/// # Panics
/// Panics if `ty` is [`AnyTypeEnum::FunctionType`].
#[must_use]
pub fn create_fn<'ctx: 'a, 'a>(
    ctx: &impl AsCompilationUnitCtx<'ctx, 'a>,
    ty: AnyTypeEnum<'ctx>,
    dbg_ty: DIType<'ctx>,
    args: &[BasicMetadataTypeEnum<'ctx>],
    dbg_args: &[DIType<'ctx>],
    is_variadic: bool,
) -> (
    FunctionType<'ctx>,
    DISubroutineType<'ctx>,
    DIBasicType<'ctx>,
) {
    (
        match ty {
            AnyTypeEnum::ArrayType(x) => x.fn_type(args, is_variadic),
            AnyTypeEnum::FloatType(x) => x.fn_type(args, is_variadic),
            AnyTypeEnum::IntType(x) => x.fn_type(args, is_variadic),
            AnyTypeEnum::PointerType(x) => x.fn_type(args, is_variadic),
            AnyTypeEnum::StructType(x) => x.fn_type(args, is_variadic),
            AnyTypeEnum::VectorType(x) => x.fn_type(args, is_variadic),
            AnyTypeEnum::VoidType(x) => x.fn_type(args, is_variadic),
            AnyTypeEnum::FunctionType(_) => panic!("fn is not a valid return type for a function"),
        },
        ctx.dbg_builder().create_subroutine_type(
            ctx.compilation_unit().get_file(),
            Some(dbg_ty),
            dbg_args,
            0,
        ),
        ctx.dbg_builder()
            .create_basic_type(&ty.to_string(), 0, 0, 0)
            .expect("basic type should be valid"),
    )
}

/// Resolve a [`Type`] to a LLVM [`IntType`]
///
/// # Panics
/// Panics if `ty` is not an integer type
#[allow(clippy::needless_pass_by_value)]
pub fn llvm_int_type<'ctx: 'a, 'a>(
    ctx: &impl AsCompilationUnitCtx<'ctx, 'a>,
    ty: &Type,
) -> (IntType<'ctx>, DIBasicType<'ctx>) {
    (
        match ty {
            Type::Bool => ctx.ctx().bool_type(),
            Type::I8 | Type::U8 => ctx.ctx().i8_type(),
            Type::I16 | Type::U16 => ctx.ctx().i16_type(),
            Type::I32 | Type::U32 => ctx.ctx().i32_type(),
            Type::I64 | Type::U64 => ctx.ctx().i64_type(),
            Type::Usize | Type::Isize => ctx.ctx().ptr_sized_int_type(
                &ctx.target_machine().get_target_data(),
                Some(AddressSpace::default()),
            ),
            Type::Int => {
                panic!("{{int}} type reached code generation, should be resolved in typeck")
            }
            Type::Ptr(_) | Type::Fn(_) | Type::Struct(_) | Type::Union(_) => {
                panic!("not an integer type")
            }
            Type::Opaque(name) => {
                panic!("opaque type '{name}' reached code generation, should be resolved in typeck")
            }
        },
        ctx.dbg_builder()
            .create_basic_type(&ty.to_string(), 0, 0, 0)
            .expect("basic type should be valid"),
    )
}

/// Resolve a [`Type`] to a LLVM [`BasicTypeEnum`]
///
/// # Panics
/// Panics if `ty` is not a basic type
#[allow(clippy::too_many_lines)]
pub fn llvm_basic_type<'ctx: 'a, 'a>(
    ctx: &impl AsCompilationUnitCtx<'ctx, 'a>,
    ty: &Type,
) -> (BasicTypeEnum<'ctx>, DIType<'ctx>) {
    match ty {
        Type::Bool
        | Type::I8
        | Type::U8
        | Type::I16
        | Type::U16
        | Type::I32
        | Type::U32
        | Type::I64
        | Type::U64
        | Type::Usize
        | Type::Isize => {
            let (ty, dbg_ty) = llvm_int_type(ctx, ty);
            (ty.as_basic_type_enum(), dbg_ty.as_type())
        }
        Type::Int => {
            panic!("{{int}} type reached code generation, should be resolved in typeck")
        }
        Type::Ptr(x) => {
            let (pointee_ty, pointee_dbg_ty) = llvm_type(ctx, x);
            let (ty, dbg_ty) = create_ptr(ctx, pointee_ty, pointee_dbg_ty);
            (ty.as_basic_type_enum(), dbg_ty.as_type())
        }
        Type::Fn(_) => panic!("function is not a basic type"),
        Type::Opaque(name) => {
            panic!("opaque type '{name}' reached code generation, should be resolved in typeck")
        }
        Type::Struct(fields) => (
            ctx.ctx()
                .struct_type(
                    &fields
                        .into_iter()
                        .map(|(_, key_ty)| llvm_basic_type(ctx, key_ty).0)
                        .collect::<Vec<_>>(),
                    false,
                )
                .as_basic_type_enum(),
            ctx.dbg_builder()
                .create_struct_type(
                    ctx.compilation_unit().get_file().as_debug_info_scope(),
                    &ty.to_string(),
                    ctx.compilation_unit().get_file(),
                    0,
                    0,
                    0,
                    0,
                    None,
                    &fields
                        .into_iter()
                        .map(|(key, key_ty)| {
                            ctx.dbg_builder()
                                .create_member_type(
                                    ctx.compilation_unit().get_file().as_debug_info_scope(),
                                    key,
                                    ctx.compilation_unit().get_file(),
                                    0,
                                    0,
                                    0,
                                    0,
                                    0,
                                    llvm_basic_type(ctx, key_ty).1,
                                )
                                .as_type()
                        })
                        .collect::<Vec<_>>(),
                    0,
                    None,
                    "",
                )
                .as_type(),
        ),
        Type::Union(fields) => {
            // Determine which field has the largest size. This is what we will allocate.
            let largest_field = fields
                .into_iter()
                .map(|(_, ty)| {
                    let ty = llvm_basic_type(ctx, ty).0;
                    let size = ctx.target_machine().get_target_data().get_bit_size(&ty);
                    (ty, size)
                })
                .max_by_key(|(_, size)| *size)
                .unwrap_or_else(|| {
                    // this is basically `never`
                    let ty = ctx.ctx().struct_type(&[], false).as_basic_type_enum();
                    (ty, ctx.target_machine().get_target_data().get_bit_size(&ty))
                });

            (
                largest_field.0,
                ctx.dbg_builder()
                    .create_union_type(
                        ctx.compilation_unit().get_file().as_debug_info_scope(),
                        &ty.to_string(),
                        ctx.compilation_unit().get_file(),
                        0,
                        0,
                        0,
                        0,
                        &fields
                            .into_iter()
                            .map(|(_, ty)| llvm_basic_type(ctx, ty).1)
                            .collect::<Vec<_>>(),
                        0,
                        "",
                    )
                    .as_type(),
            )
        }
    }
}

/// Resolve a [`Type`] to a LLVM [`AnyTypeEnum`]
pub fn llvm_type<'ctx: 'a, 'a>(
    ctx: &impl AsCompilationUnitCtx<'ctx, 'a>,
    ty: &Type,
) -> (AnyTypeEnum<'ctx>, DIType<'ctx>) {
    match ty {
        Type::Bool
        | Type::I8
        | Type::U8
        | Type::I16
        | Type::U16
        | Type::I32
        | Type::U32
        | Type::I64
        | Type::U64
        | Type::Usize
        | Type::Isize
        | Type::Ptr(_)
        | Type::Struct(_)
        | Type::Union(_) => {
            let (ty, dbg_ty) = llvm_basic_type(ctx, ty);
            (ty.as_any_type_enum(), dbg_ty)
        }
        Type::Int => {
            panic!("{{int}} type reached code generation, should be resolved in typeck")
        }

        Type::Fn(Fn { arguments, returns }) => {
            let (ret, ret_dbg) = llvm_type(ctx, returns);
            let is_variadic = arguments.is_variadic();
            let (fn_ty, _, fn_dbg_ty) = create_fn(
                ctx,
                ret,
                ret_dbg,
                &arguments
                    .as_arguments()
                    .iter()
                    .map(|arg| llvm_basic_type(ctx, arg.ty.value()).0.into())
                    .collect::<Vec<_>>(),
                &arguments
                    .as_arguments()
                    .iter()
                    .map(|arg| llvm_basic_type(ctx, arg.ty.value()).1)
                    .collect::<Vec<_>>(),
                is_variadic,
            );
            (fn_ty.as_any_type_enum(), fn_dbg_ty.as_type())
        }

        Type::Opaque(name) => {
            panic!("opaque type '{name}' reached code generation, should be resolved in typeck")
        }
    }
}

#[cfg(test)]
mod tests {
    // Please read the "Common patterns in tests" section of crate::test_utils for
    // more information on how code generator tests are structured.

    use indoc::indoc;

    use crate::cg_snapshot_test;

    #[test]
    fn tagged_unions_are_properly_typed() {
        cg_snapshot_test!(indoc! {"
            enum VariableInt {
                Eight: i8,
                Sixteen: i16,
            }

            fn test() -> VariableInt {
                // TEST: should generate a { usize, i16 (largest of values) }
                let x: VariableInt;

                // internal parameters
                let y = x.__discriminant__;
                let z = x.__value__;

                return x;
            }
        "});
    }
}
