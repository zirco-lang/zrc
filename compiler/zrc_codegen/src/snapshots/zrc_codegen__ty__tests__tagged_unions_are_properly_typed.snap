---
source: compiler/zrc_codegen/src/ty.rs
description: "enum VariableInt {\n    Eight: i8,\n    Sixteen: i16,\n}\n\nfn test() -> VariableInt {\n    // TEST: should generate a { usize, i16 (largest of values) }\n    let x: VariableInt;\n\n    // internal parameters\n    let y = x.__discriminant__;\n    let z = x.__value__;\n\n    return x;\n}\n"
expression: resulting_ir
---
; ModuleID = 'test.zr'
source_filename = "test.zr"

define { i64, i16 } @test() !dbg !3 {
entry:
  %let_z = alloca i16, align 2
  %let_y = alloca i64, align 8
  %let_x = alloca { i64, i16 }, align 8
  call void @llvm.dbg.declare(metadata ptr %let_x, metadata !16, metadata !DIExpression()), !dbg !21
  call void @llvm.dbg.declare(metadata ptr %let_y, metadata !19, metadata !DIExpression()), !dbg !22
  %gep = getelementptr inbounds { i64, i16 }, ptr %let_x, i32 0, i32 0, !dbg !23
  %load = load i64, ptr %gep, align 4, !dbg !23
  store i64 %load, ptr %let_y, align 4, !dbg !22
  call void @llvm.dbg.declare(metadata ptr %let_z, metadata !20, metadata !DIExpression()), !dbg !24
  %gep1 = getelementptr inbounds { i64, i16 }, ptr %let_x, i32 0, i32 1, !dbg !25
  %load2 = load i16, ptr %gep1, align 2, !dbg !25
  store i16 %load2, ptr %let_z, align 2, !dbg !24
  %load3 = load { i64, i16 }, ptr %let_x, align 4, !dbg !26
  ret { i64, i16 } %load3, !dbg !26
}

; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
declare void @llvm.dbg.declare(metadata, metadata, metadata) #0

attributes #0 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }

!llvm.module.flags = !{!0}
!llvm.dbg.cu = !{!1}

!0 = !{i32 2, !"Debug Info Version", i32 3}
!1 = distinct !DICompileUnit(language: DW_LANG_C, file: !2, producer: "zrc test runner", isOptimized: false, flags: "zrc --fake-args", runtimeVersion: 0, emissionKind: FullDebug, splitDebugInlining: false)
!2 = !DIFile(filename: "test.zr", directory: "/fake/path")
!3 = distinct !DISubprogram(name: "test", linkageName: "test", scope: null, file: !2, line: 6, type: !4, scopeLine: 6, spFlags: DISPFlagDefinition, unit: !1, retainedNodes: !15)
!4 = !DISubroutineType(types: !5)
!5 = !{!6}
!6 = !DICompositeType(tag: DW_TAG_structure_type, name: "struct { __discriminant__: usize, __value__: union { Eight: i8, Sixteen: i16 } }", scope: !2, file: !2, elements: !7)
!7 = !{!8, !10}
!8 = !DIDerivedType(tag: DW_TAG_member, name: "__discriminant__", scope: !2, file: !2, baseType: !9)
!9 = !DIBasicType(name: "usize")
!10 = !DIDerivedType(tag: DW_TAG_member, name: "__value__", scope: !2, file: !2, baseType: !11)
!11 = !DICompositeType(tag: DW_TAG_union_type, name: "union { Eight: i8, Sixteen: i16 }", scope: !2, file: !2, elements: !12)
!12 = !{!13, !14}
!13 = !DIBasicType(name: "i8")
!14 = !DIBasicType(name: "i16")
!15 = !{!16, !19, !20}
!16 = !DILocalVariable(name: "x", scope: !17, file: !2, line: 8, type: !6)
!17 = distinct !DILexicalBlock(scope: !18, file: !2, line: 6, column: 26)
!18 = distinct !DILexicalBlock(scope: !3, file: !2, line: 6, column: 26)
!19 = !DILocalVariable(name: "y", scope: !17, file: !2, line: 11, type: !9)
!20 = !DILocalVariable(name: "z", scope: !17, file: !2, line: 12, type: !11)
!21 = !DILocation(line: 8, column: 9, scope: !17)
!22 = !DILocation(line: 11, column: 9, scope: !17)
!23 = !DILocation(line: 11, column: 13, scope: !17)
!24 = !DILocation(line: 12, column: 9, scope: !17)
!25 = !DILocation(line: 12, column: 13, scope: !17)
!26 = !DILocation(line: 14, column: 12, scope: !17)
